File: tree-inorder-preorder.py
'''
Time Complexity : O(n)
Space Complexity: O(n)


'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        # inorder L root Right
        # postorder L right Root
        # preorder Root Left Right
        
        # Need to see solution to understand
        
        # root will be first of preorder
        inord = {}
        for k,v in enumerate(inorder):
            inord[v] = k
        preord = deque(preorder)
        i=0
        def build(preorder,start,end):
            
            if(start>=end):
                return
            # root
            
            # can perform popleft or use a variable i in global scope
            #root_val = preorder.popleft()
            
            nonlocal i
            root_val = preorder[i]
            root = TreeNode(root_val)
            subtree_ind  = inord[root_val]
            i+=1
            # left subtree
            root.left = build(preorder,start, subtree_ind)
            # right subtree
            root.right = build(preorder,subtree_ind+1, end)
            
            return root
        return build(preord, 0, len(preord))

File: treeInorderPreorder.cpp
// Time Complexity :O(n) where n in the number of nodes
// Space Complexity : O(n)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int l = 0;
        return helper(l,0,inorder.size()-1,preorder,inorder);
    }
    TreeNode* helper(int &preStart,int inStart,int inEnd,vector<int>& preorder,vector<int>& inorder){
        if(inStart > inEnd || preStart >= preorder.size()) return nullptr;
        TreeNode *root = new TreeNode(preorder[preStart]);
        int inIndex;
        for(int i = inStart;i<=inEnd;i++){
            if(inorder[i] == preorder[preStart])
                inIndex =i;
        }
        preStart++;
        root->left =helper(preStart,inStart,inIndex-1,preorder,inorder);
        root->right =helper(preStart,inIndex+1,inEnd,preorder,inorder);
        return root;
    }
};